zebrad/src/components/sync/tests/vectors.rs

```diff
@@ -0,0 +1,128 @@
+    );
+
+    // This constraint avoids spurious failures after checkpointing has finished
+    assert!(
+        BLOCK_VERIFY_TIMEOUT.as_secs()
+            > 2 * zebra_chain::parameters::NetworkUpgrade::Blossom
+                .target_spacing()
+                .num_seconds() as u64,
+        "Block verify should allow for at least one new block to be generated and distributed"
+    );
+}
+
+/// Test that calls to [`ChainSync::request_genesis`] are rate limited.
+#[test]
+fn request_genesis_is_rate_limited() {
+    // create some counters that will be updated inside async blocks
+    let peer_requests_counter = Arc::new(AtomicU8::new(0));
+    let peer_requests_counter_in_service = Arc::clone(&peer_requests_counter);
+    let state_requests_counter = Arc::new(AtomicU8::new(0));
+    let state_requests_counter_in_service = Arc::clone(&state_requests_counter);
+
+    let runtime = Runtime::new().expect("Failed to create Tokio runtime");
+    let _guard = runtime.enter();
+
+    // create a fake peer service that respond with `Error` to `BlocksByHash` or
+    // panic in any other type of request.
+    let peer_service = tower::service_fn(move |request| {
+        match request {
+            zebra_network::Request::BlocksByHash(_) => {
+                // Track the call
+                peer_requests_counter_in_service.fetch_add(1, Ordering::SeqCst);
+                // Respond with `Error`
+                future::err("block not found".into())
+            }
+            _ => unreachable!("no other request is allowed"),
+        }
+    });
+
+    // create a state service that respond with `None` to `Depth` or
+    // panic in any other type of request.
+    let state_service = tower::service_fn(move |request| {
+        match request {
+            zebra_state::Request::Depth(_) => {
+                // Track the call
+                state_requests_counter_in_service.fetch_add(1, Ordering::SeqCst);
+                // Respond with `None`
+                future::ok(zebra_state::Response::Depth(None))
+            }
+            _ => unreachable!("no other request is allowed"),
+        }
+    });
+
+    // create a verifier service that will always panic as it will never be called
+    let verifier_service =
+        tower::service_fn(
+            move |_| async move { unreachable!("no request to this service is allowed") },
+        );
+
+    // start the sync
+    let mut chain_sync = ChainSync::new(
+        &ZebradConfig::default(),
+        peer_service,
+        state_service,
+        verifier_service,
+    );
+
+    // run `request_genesis()` with a timeout of 13 seconds
+    runtime.block_on(async move {
+        let _ = timeout(Duration::from_secs(13), chain_sync.request_genesis()).await;
+    });
+
+    // In 13 seconds and as we have a rate limit of 5 seconds in `request_genesis()` ...
+
+    // we should have 3 calls to the peer service
+    assert_eq!(peer_requests_counter.load(Ordering::SeqCst), 9);
```

Comment:
Note to self: research why 9.

Reply:
The zebra sync service has two kinds of retries, both of them can double the number of requests. So the number of requests can be up to 4x the number of times you call `request_genesis`.

Also lets use a constant for the number of calls to `request_genesis`. You can see an example in:
https://github.com/ZcashFoundation/zebra/pull/2254/files#diff-8853e569e32adaf7d920a17860992e7076b11d376a2844aeabe397fe28e343d8R140

```suggestion
    let peer_requests_counter = peer_requests_counter.load(Ordering::SeqCst);
    assert!(peer_requests_counter >= RETRIES_TO_RUN);
    assert!(peer_requests_counter <= RETRIES_TO_RUN * BLOCK_DOWNLOAD_RETRY_LIMIT * 2);
```

Here are the details for the block download retries:
https://github.com/ZcashFoundation/zebra/blob/main/zebrad/src/components/sync.rs#L29-L37